src/js/client/templateAware.js

```diff
@@ -0,0 +1,124 @@
+
+     renderInitialMarkup: {
+        func: "{that}.renderMarkup",
+        args: [
+          "{that}",
+          "{that}.options.selectors.selector",
+          "{that}.options.templates.templates",
+          "{that}.model",
+          "appendTo"
+        ]
+      }
+
+  For an example of using this in depth, check out the provided `templateFormControl` grade or the client side tests.
+ */
+/* global fluid, jQuery */
+(function () {
+    "use strict";
+    var gpii = fluid.registerNamespace("gpii");
+    fluid.registerNamespace("gpii.templates.templateAware");
+
+    // A convenience function that can be used to more easily define `renderInitialMarkup` invokers (see example above).
+    gpii.templates.templateAware.renderMarkup = function (that, renderer, selector, template, data, manipulator) {
+        manipulator = manipulator ? manipulator : "html";
+        var element = that.locate(selector);
+        if (renderer) {
+            renderer[manipulator](element, template, data);
+            that.events.onMarkupRendered.fire(that);
+        }
+        else {
+            fluid.fail("I cannot render content without a renderer.");
+        }
+    };
+
+    // When overriding this, you should fire an `onMarkupRendered` event to ensure that bindings can be applied.
+    gpii.templates.templateAware.noRenderFunctionDefined = function () {
+        fluid.fail("You are expected to define a renderInitialMarkup invoker when implementing a templateAware component.");
+    };
+
+    gpii.templates.templateAware.refreshDom = function (that) {
+        // Adapted from: https://github.com/fluid-project/infusion/blob/master/src/framework/preferences/js/Panels.js#L147
+        var userJQuery = that.container.constructor;
+        that.container = userJQuery(that.container.selector, that.container.context);
+        fluid.initDomBinder(that, that.options.selectors);
+        that.events.onDomBind.fire(that);
+    };
+
+    fluid.defaults("gpii.templates.templateAware", {
+        gradeNames: ["fluid.viewRelayComponent", "gpii.hasRequiredOptions", "autoInit"],
+        requiredOptions: {
+            templates:           true,
+            "templates.initial": true,
+            "templates.error":   true,
+            "templates.success": true
+        },
+        events: {
+            refresh: null,
+            onMarkupRendered: null,
+            onDomBind: null
+        },
+        listeners: {
+            "refresh.renderMarkup": {
+                func: "{that}.renderInitialMarkup"
+            },
+            "onDomBind.applyBinding": {
+                funcName: "gpii.templates.binder.applyBinding",
+                args:     ["{that}"]
+            },
+            "onMarkupRendered.refreshDom": {
+                funcName: "gpii.templates.templateAware.refreshDom",
+                args:     ["{that}"]
+            }
+        },
+        invokers: {
+            noRenderFunctionDefined: {
```

Comment:
No need for a separate invoker here - this isn't usefully configurable behaviour. You can just bind directly to funcName: "gpii.templates.templateAware.noRenderFunctionDefined" in renderInitialMarkup. Please also leave a note to upgrade to the framework facility coming in https://issues.fluidproject.org/browse/FLUID-5733 when the framework is updated.